python/openai/tests/test_tool_calling.py

```diff
@@ -0,0 +1,571 @@
+        role_name: Optional[str] = None
+        finish_reason_count: int = 0
+
+        # make the same request, streaming
+        stream = await client.chat.completions.create(
+            model=model,
+            messages=MESSAGES_ASKING_FOR_TOOLS,
+            temperature=0,
+            max_tokens=128,
+            tools=[WEATHER_TOOL, WEATHER_FORECAST_TOOL],
+            logprobs=False,
+            stream=True,
+        )
+
+        async for chunk in stream:
+            assert chunk.choices[0].index == 0
+
+            if chunk.choices[0].finish_reason:
+                finish_reason_count += 1
+                assert chunk.choices[0].finish_reason == "tool_calls"
+
+            # if a role is being streamed make sure it wasn't already set to
+            # something else
+            if chunk.choices[0].delta.role:
+                assert not role_name or role_name == "assistant"
+                role_name = "assistant"
+
+            # if a tool call is streamed make sure there's exactly one
+            # (based on the request parameters
+            streamed_tool_calls = chunk.choices[0].delta.tool_calls
+
+            if streamed_tool_calls and len(streamed_tool_calls) > 0:
+                assert len(streamed_tool_calls) == 1
+                tool_call = streamed_tool_calls[0]
+
+                # if a tool call ID is streamed, make sure one hasn't been already
+                if tool_call.id:
+                    assert not tool_call_id
+                    tool_call_id = tool_call.id
+
+                # if parts of the function start being streamed
+                if tool_call.function:
+                    # if the function name is defined, set it. it should be streamed
+                    # IN ENTIRETY, exactly one time.
+                    if tool_call.function.name:
+                        assert function_name is None
+                        assert isinstance(tool_call.function.name, str)
+                        function_name = tool_call.function.name
+                    if tool_call.function.arguments:
+                        assert isinstance(tool_call.function.arguments, str)
+                        function_args_str += tool_call.function.arguments
+
+        assert finish_reason_count == 1
+        assert role_name == "assistant"
+        assert isinstance(tool_call_id, str) and (len(tool_call_id) >= 9)
+
+        # validate the name and arguments
+        assert function_name == WEATHER_TOOL["function"]["name"]
+        assert function_name == tool_calls[0].function.name
+        assert isinstance(function_args_str, str)
+
+        # validate arguments
+        streamed_args = json.loads(function_args_str)
+        assert isinstance(streamed_args, Dict)
+        assert isinstance(streamed_args.get("city"), str)
+        assert isinstance(streamed_args.get("state"), str)
+        assert streamed_args.get("city") == "Dallas"
+        assert streamed_args.get("state") == "TX"
```

Comment:
I think this whole assertion chunk is copied a few times, and can probably be moved to a helper to reduce code dupe (feel free to change the function name to what makes most sense):
```python
self.validate_weather_tool_outputs()
```